src/js/sonifier.js

```diff
@@ -0,0 +1,402 @@
+                gradeNames: ["floe.chartAuthoring.electricPianoBand"],
+
+                components: {
+                    scheduler: {
+                        type: "flock.scheduler.async",
+                        options: {
+                            synthContext: "floe.chartAuthoring.electricPianoBand"
+                        }
+                    }
+                }
+            });
+
+            var dataPianoBand = floe.chartAuthoring.dataPianoBand();
+
+            // Add the synth to the model
+            that.applier.change("synth", dataPianoBand);
+
+            var sonifiedData = that.model.sonifiedData;
+            // Copy the sonification definition into the queue
+            that.applier.change("sonificationQueue",sonifiedData);
+        }
+    };
+
+    // Passed a sonified dataset, this function + playDataAndQueueNext acts recursively
+    // to loop through the dataset, play a voice label + sonified data, then
+    // schedule the next play event based on the timing
+
+    // The "delay" variable manages the time to elapse (in seconds) before
+    // beginning the voice label + sonification of the particular datapoint.
+    // This is required for scheduling with the Flocking-based synth generator,
+    // which plays asynchronously without callbacks, while voice events have
+    // callbacks to indicate their speech has completed
+    //
+    // The basic challenge of scheduling that this approach handles is:
+    // - we can calculate in advance how long the sonification of a data point
+    // takes by summing note duration, but we can't receive event notification
+    // when a sonification completes
+    // - we can fire an event when a voice label read completes, but can't know
+    // in advance how long it will take to read the label
+    floe.chartAuthoring.sonifier.processSonificationQueue = function(delay, noGap, that) {
+        that.applier.change("isPlaying", true);
+        var gapDuration = that.options.playbackOptions.gapDuration;
+
+        var sonificationQueue = that.model.sonificationQueue;
+
+        var synth = that.model.synth;
+
+        var currentData = sonificationQueue[0];
+
+        var textToSpeech = that.textToSpeech;
+
+        // Schedule the next voice label, accounting for both the variable-length
+        // delay (the time to play the preceding sonification) and the fixed-length
+        // gap
+
+        // We shouldn't use the gap if this is the first call for a dataset
+        var pause = noGap? delay : delay+gapDuration;
+        if(fluid.textToSpeech.isSupported()) {
+            synth.scheduler.once(pause, function() {
+                that.applier.change("currentlyPlayingData", currentData);
+                textToSpeech.queueSpeech(currentData.label);
+            });
+        }
+    };
+
+    // Recursion function called from floe.chartAuthoring.sonifier.processSonificationQueue
+
+    floe.chartAuthoring.sonifier.playDataAndQueueNext = function(that) {
```

Comment:
There is a some rudimentary capacity to declaratively schedule things using Flocking's scheduler, which was implemented hastily in a motel room in Boulder a few years ago. However, it's wildly insufficient, especially for this kind of "mixed scheduling" where different types of actions (scheduling changes on a synth, playing back speech, etc.) are being interleaved. Rationalizing this kind of thing in a fully declarative form is a job for BERGSON 2.0!

"There is no state of mind, however simple, which does not change every moment, since there is no consciousness without memory, and no continuation of a state without the addition, to the present feeling, of the memory of past moments. It is this which constitutes duration."